### PR TITLE
fix: [DVM2.0] deleting rolled spam requests

### PR DESCRIPTION
**Pull request by Reinis-FRP**
**Motivation**

After experimenting with the combination of rolled votes + spam deletion requests we identified a subtle issue if you try and delete a rolled spam vote wherein the voting contract would: a) not correctly update the jump mapping and b) re-euqnue spam deletion requests within the pending requests array.

To address this, this PR adds a small logic change and additional unit tests to better accommodate spam deletion + rolled votes. This is critical as in the event that spam is created and UMA governance is too slow to initiate a deletion _before_ voting begins on the spam (i.e dont submit the spam deletion request within the [minRollToNextRoundLength](https://github.com/UMAprotocol/protocol/blob/25d22a9088a64da8f6783130c65d717a9104015d/packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol#L233) period) we could hit a situation quite easily where we need to first _roll_ spam and _then_ delete it. In practice, this could look something like this:
1) right before cycle C0 ends, and before `phaseEndTime-minRollToNextRoundLength`, a spammer submits a large number of requests.
2) governance misses the chance to signal for these to be deleted in the next round (within the `minRollToNextRoundLength`). this could also happen if `minRollToNextRoundLength` is set to 0.
3) the next cycle C1 starts. now, The spam votes are available to be voted on. However, UMA voters _know_ these are spam and dont want to vote on them. During C1 someone submits a spam deletion request to remove these from the voting system.
4) the voting round ends without any spam meeting the GAT.
5) Now, we are in C2. The spam was enqued in C0, not settled in C1 and is now voted on again in C2, along with the spam deletion request. This amounts to the spam rolling. The voting contracts should enable voters to _only_ vote on the spam deletion request in C2, without voting on any of the spam, and then in C3 when the spam deletion request settles and is executed all spam should be removed from the contracts.
6) Critically at this point **no voters should need to traversed by any other voters as the spam was removed**. I.e the cost of using the voting system (and updating trackers) should not be affected at all removal of the spam, even though it was rolled.

### CODE REVIEW

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 1]
```diff
@@ -1006,33 +1010,48 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
      */
 
     function executeSpamDeletion(uint256 proposalId) external nonReentrant() {
-        require(spamDeletionProposals[proposalId].executed == false, "Proposal already executed");
```
Comment (chrismaree) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1009]: we hit the contract size limit so I removed some requires that will never really be hit.

File: packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol [hunk 2]
```diff
@@ -1006,33 +1010,48 @@ contract VotingV2 is Staker, OracleInterface, OracleAncillaryInterface, OracleGo
...
 
         bytes32 identifier = SpamGuardIdentifierLib._constructIdentifier(SafeCast.toUint32(proposalId));
 
         (bool hasPrice, int256 resolutionPrice, ) =
             _getPriceOrError(identifier, spamDeletionProposals[proposalId].requestTime, "");
-        require(hasPrice, "Spam proposal has not resolved");
+        require(hasPrice);
 
         // If the price is non zero then the spam deletion request was voted up to delete the requests. Execute delete.
         if (resolutionPrice != 0) {
```
Comment (chrismaree) [inline packages/core/contracts/data-verification-mechanism/implementation/VotingV2.sol:1023]: the core logic can be seen below: if we detect that a vote was rolled then we need to store the range over which the roll occurred and used this to update the `skippedRequestIndexes` to accommodate this.

**Review by mrice32 — COMMENTED** [review]
Wow, this is an interesting find. This was found _outside of the audit_, right?

I'm assuming this fix is no longer relevant given that we're removing the jump logic. Does this mean we should _only_ add the tests from this fix once the refactor is done?

**Comment by chrismaree** [comment]
This PR is now redundant and is closed in favor of: https://github.com/UMAprotocol/protocol/pull/4352

**Pull request closed without merging**